#166272345 Users should like and unlike Articles

adex001 opened the pull request:
#### What does this PR do?
- Users should be able to like and unlike an article
#### Description of Task to be completed?
- Create a like and unlike endpoint
- Verify the endpoints to give access to only valid users
- Write the like and unlike functionalities
- Write unit test
#### How should this be manually tested?
Navigate to the endpoint `POST {{HOST_NAME}}/api/v1/articles/:slug/like` and `DELETE {HOST_NAME}}/api/v1/articles/:slug/like`
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166272345](https://www.pivotaltracker.com/story/show/166272345)
#### Screenshots (if appropriate)
<img width="1440" alt="Screenshot 2019-06-19 at 10 28 48 AM" src="https://user-images.githubusercontent.com/29796551/59754442-03df2100-927e-11e9-9280-229844bbacab.png">

#### Questions:
N/A

adex001 committed:
feat-[166272346]: Users follow each other
create a follow endpoint
verify the endpoint to allow only valid users
write the follow functionality
make sure users cannot follow themselves
make sure a user cannot follow another user multiple times
update the routes to be version compliant

adex001 committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272346-ft-users-follow-each-other

adex001 committed:
feat-[166272346]: Users follow each other
update and write more tests

adex001 committed:
feat-[166272346]: Users follow each other
update and write more test
pull from develop and resolve merge conflict
used faker to mock up datas
change test to use Bearer <token> as authorization

adex001 committed:
feat-[166272346]: Users follow each other
update feedback from PR

adex001 committed:
feat-[166272346]: Users follow each other
update feedback from PR and update tests

adex001 committed:
feat-[166272346]: Users follow each other
update merge conflict by pulling from develop
update endpoint to fit in documentation
write unfollow functionality
write unit tests for unfollow endpoint

adex001 committed:
feat-[166272346]: Users follow each other
update merge conflict by pulling from develop
update endpoint to fit in documentation
write unfollow functionality
write unit tests for unfollow endpoint

adex001 committed:
all

adex001 committed:
update feedback

adex001 committed:
Update the follow and unfollow using magic method

adex001 committed:
feat-[166272346]: Users follow each other
update feedback

adex001 committed:
feat-[166272346]: Users follow each other
updated feedback by using req.user in the follow controller
updated route index.js

adex001 committed:
feat-[166272346]: Users follow each other
resolve merge conflicts

adex001 committed:
feat-[166272346]: Users follow each other
add the follow endpoint to return a profile for  both follow anf unfolloe

adex001 committed:
feat-[166272346]: Users follow each other
add the create-follower file back to migrations

adex001 committed:
feat-[166272346]: Users follow each other
move the follow and unfollow controller from user controller to profile controller
update the follow and unfollow method to return profile of who followed

adex001 committed:
feat-[166272346]: Users follow each other
move the follow and unfollow controller from user controller to profile controller
update the follow and unfollow method to return profile of who followed

adex001 committed:
feat-[166272346]: Users follow each other
move the follow and unfollow controller from user controller to profile controller
update the follow and unfollow method to return profile of who followed

adex001 committed:
feat-[166272346]: Users follow each other
resolved merge conflicts

adex001 committed:
feat-[166272346]: Users follow each other
resolved merge conflicts

adex001 committed:
feat-[166272346]: Users follow each other
add more tests for updateProfile when there is a validation error

adex001 committed:
feat-[166272346]: Users follow each other
return an object in the optional payload in both follow and unfollow endpoint

adex001 committed:
Merge branch 'develop' of https://github.com/andela/kingsmen-ah-backend into 166272345-ft-article-like-dislike

adex001 committed:
feat-[166272345-ft-article-like-dislike
delete unneccessary routes/index.js file

adex001 committed:
feat-[166272345-ft-article-like-dislike
implemet follow and unfollow feature

adex001 committed:
feat-[166272345-ft-article-like-dislike
write tests for like and unlike functiomalities

adex001 committed:
feat-[166272345-ft-article-like-dislike
fix merge conflict

adex001 committed:
feat-[166272345-ft-article-like-dislike
update swagger documentation

adex001 committed:
feat-[166272345-ft-article-like-dislike
increase test coverage

adex001 committed:
feat-[166272345-: Users can like or unlike articles
increase test coverage

adex001 committed:
feat-[166272345]: Users can like or unlike articles
- increase test coverage
[Delivers #166272345]

adex001 committed:
feat-[166272345]: Implement exactly once delivery  
- increase test coverage  
[Delivers #166272345]

adex001 committed:
feat-[166272345]: Users like and unlike an article
- increase test coverage
[Delivers #166272345]

adex001 committed:
feat-[166272345]: Users like and unlike

- increase test coverage
- like and unlike an article

[Finishes #166272345]

despeauxz reviewed (dismissed):
Looks good

fidelisojeah commented on server/routes/api/articles.js:
```diff
@@ -19,5 +19,9 @@ articlesRouter.get('/:slug', ArticleController.getOne);
 articlesRouter.post('/', Token.authorize, ArticleController.create);
 articlesRouter.put('/:slug', Token.authorize, ArticleController.update);
 articlesRouter.delete('/:slug', Token.authorize, ArticleController.delete);
+articlesRouter.post('/:slug/rate', trim, Token.authorize, ArticleController.rate);
```
already appears on line 11
```suggestion

```

adex001 replied:
okay. Thanks.

adex001 committed:
feat-[166272345]: Implement Article like and unlike

- remove duplicate rateArticle endpoint

[Finishes #166272345]

emmygozi commented on server/controllers/articles.js:
```diff
@@ -214,6 +214,52 @@ class ArticleController {
+  */
+  static async like(req, res, next) {
+    try {
+      const me = req.user;
```
Consider changing variable name ```me``` to something more descriptive to avoid confusion and explain better.
Consider using a phrase in camel casing.

Looks great to me otherwise.

adex001 replied:
Thanks for your feedback but using me is descriptive enough. It simply refers to **the person** making the request.

fidelisojeah commented on server/controllers/articles.js:
```diff
@@ -214,6 +214,52 @@ class ArticleController {
+    try {
+      const me = req.user;
+      const { slug } = req.params;
+      const article = await findArticle({ slug });
```
you should include the number of likes an article has in its attribute.

Please modify the `findArticle` method to include the sum of likes an article has.

Please only the sum not details

adex001 replied:
done and dusted.

adex001 committed:
feat-[166272345]: article like dislike

implement total number of likes in the findArticle method

[Finishes #166272345]

adex001 committed:
feat-[166272345]: article like dislike

- resolve merge conflicts

[Finishes #166272345]

adex001 committed:
feat-[166272345]: article like dislike

- update test to increase coverage

[Finishes #166272345]

adex001 committed:
feat-[166272345]: article like dislike

- update test to increase coverage

[Finishes #166272345]

GerrardE reviewed (dismissed):
I like the fact that you carefully documented this feature.

adex001 committed:
feat-[166272345]: article like dislike

- update test to increase coverage

[Finishes #166272345]

fidelisojeah commented on swagger.yaml:
```diff
@@ -867,4 +829,4 @@ definitions:
 externalDocs:
   description: "Find out more about Author's Haven"
-  url: "https://github.com/andela/kingsmen-ah-backend"
+  url: "https://github.com/andela/kingsmen-ah-backend"
```
new line?

fidelisojeah commented on server/migrations/20190619081247-likearticle.js:
```diff
@@ -0,0 +1,21 @@
+module.exports = {
```
not needed file

adex001 committed:
feat-[166272345]: article like dislike

- implement feedback

[Finishes #166272345]

adex001 committed:
feat-[166272345]: article like dislike

- implement feedback

[Finishes #166272345]

adex001 committed:
feat-[166272345]: article like dislike

- fix some bugs from rating and liking

[Finishes #166272345]

emmygozi commented on server/routes/api/articles.js:
```diff
@@ -19,6 +19,10 @@ articlesRouter.delete('/:slug/comments/:id/like', authorizeAccess, find, Comment
 
 articlesRouter.get('/', ArticleController.getAll);
 articlesRouter.get('/:slug', ArticleController.getOne);
+
```
Can an unverified user like an article? If they can please let me know.

emmygozi commented on server/routes/api/articles.js:
```diff
@@ -19,6 +19,10 @@ articlesRouter.delete('/:slug/comments/:id/like', authorizeAccess, find, Comment
+
+articlesRouter.post('/:slug/like', authorizeAccess, ArticleController.like);
+articlesRouter.delete('/:slug/like', authorizeAccess, ArticleController.unlike);
+
```
Can an unverified user like an article? If they can please let me know.

adex001 replied:
Thanks for pointing this out. An unverified user should not be able to like an article.

adex001 committed:
feat-[166272345]: article like dislike

- implement Ahaiwe feedbak by making only verified user like and unlike articles

[Finishes #166272345]

GerrardE reviewed (approved):
Although the `likeCount` has been removed compared to the previous PR, you have successfully confirmed why you did so. Good Job.
